compiler/resolution/implementForallIntents.cpp

```diff
@@ -920,6 +947,96 @@ static void resolveAndPruneExplicitShadowVars(ForallStmt* fs,
   }
 }
 
+// If it is a CallExpr(<something>, _mt, arg), and <something>
+// names a field in 'recType', return that field's Symbol
+static Symbol* isFieldAccess(AggregateType* recType, Expr* arg) {
+  if (CallExpr* call = toCallExpr(arg->parentExpr))
+    if (call->numActuals() == 2)
+      if (call->get(1)->getValType() == dtMethodToken)
+        // Field accesses are unresolved method calls at this point.
+        if (UnresolvedSymExpr* base = toUnresolvedSymExpr(call->baseExpr))
+          if (Symbol* fieldSym = recType->getField(base->unresolved, false))
+            return fieldSym;
+  return NULL;
+}
+
+// Given a field "myField", create a shadow variable myField_SV
+// whose outer variable "myField_ref" is set up prior to 'fs':
+//   ref myField_ref = ovar.myField;
+static ShadowVarSymbol* createSVforFieldAccess(ForallStmt* fs, Symbol* ovar,
+                                               Symbol* field)
+{
+  bool isConst = ovar->isConstant() || field->isConstant();
+  VarSymbol* fieldRef = newTemp(astr(field->name, "_ref"),
+                                field->type->getRefType());
+  fieldRef->qual = isConst ? QUAL_CONST_REF : QUAL_REF;
+  if (isConst) fieldRef->addFlag(FLAG_CONST);
+  fs->insertBefore(new DefExpr(fieldRef));
+
+  BlockStmt* holder = new BlockStmt();
+  fs->insertBefore(holder);
+  // This is how a ref variable is currently initialized.
+  CallExpr* init = new CallExpr(field->name, gMethodToken, ovar);
+  holder->insertAtTail(new CallExpr(PRIM_MOVE, fieldRef, init));
+  resolveBlockStmt(holder);
+  holder->flattenAndRemove();
+
+  // Now create the shadow variable.
+  Type*           svarType   = field->type;
+  ForallIntentTag svarIntent = isConst ? TFI_CONST : TFI_DEFAULT;
+  bool            pruneDummy = false;
+  resolveShadowVarTypeIntent(svarType, svarIntent, pruneDummy);
+
+  ShadowVarSymbol* svar = new ShadowVarSymbol(svarIntent,
+                                              astr(field->name, "_svar"),
+                                              new SymExpr(fieldRef));
+  svar->addFlag(FLAG_DEFAULT_INTENT_IS_REF_MAYBE_CONST);//to cullOverReferences
```

Comment:
Is this copying flags from the other variable? AFAIK cull over references considers all references that don't have FLAG_REF_VAR to be ref-maybe-const. Anyway since it's not an argument and it's not on a type this seems like an abuse of FLAG_DEFAULT_INTENT_IS_REF_MAYBE_CONST. 

Could you check if you can get it working by removing FLAG_REF_VAR, if it exists, in the right place? And if not, add a different flag FLAG_REF_MAYBE_CONST for this to use and for cullOverReferences to check?